src/tests/async/synchronization-context/synchronization-context.cs

```diff
@@ -104,4 +104,70 @@ public void OnCompleted(Action continuation)
 
         public void GetResult() { }
     }
+
+    [Fact]
+    public static void TestSyncContextSaveRestore()
+    {
+        SynchronizationContext prevContext = SynchronizationContext.Current;
+        try
+        {
+            SynchronizationContext.SetSynchronizationContext(new SyncContextWithoutRestore());
+            TestSyncContextSaveRestoreAsync().GetAwaiter().GetResult();
+        }
+        finally
+        {
+            SynchronizationContext.SetSynchronizationContext(prevContext);
+        }
+    }
+
+    private static async Task TestSyncContextSaveRestoreAsync()
+    {
+        Assert.True(SynchronizationContext.Current is SyncContextWithoutRestore);
+        await ClearSyncContext();
+        Assert.True(SynchronizationContext.Current is SyncContextWithoutRestore);
+    }
+
+    private static async Task ClearSyncContext()
+    {
+        SynchronizationContext.SetSynchronizationContext(null);
+    }
+
+    [Fact]
+    public static void TestSyncContextNotRestored()
+    {
+        SynchronizationContext prevContext = SynchronizationContext.Current;
+        try
+        {
+            SynchronizationContext.SetSynchronizationContext(new SyncContextWithoutRestore());
+            TestSyncContextNotRestoredAsync().GetAwaiter().GetResult();
+        }
+        finally
+        {
+            SynchronizationContext.SetSynchronizationContext(prevContext);
+        }
+    }
+
+    private static async Task TestSyncContextNotRestoredAsync()
+    {
+        Assert.True(SynchronizationContext.Current is SyncContextWithoutRestore);
+        await SuspendThenClearSyncContext();
+        Assert.Null(SynchronizationContext.Current);
+    }
+
+    private static async Task SuspendThenClearSyncContext()
+    {
+        await Task.Yield();
+        SynchronizationContext.SetSynchronizationContext(null);
+    }
+
+    private class SyncContextWithoutRestore : SynchronizationContext
+    {
+        public override void Post(SendOrPostCallback d, object state)
+        {
+            ThreadPool.UnsafeQueueUserWorkItem(_ =>
+            {
+                d(state);
```

Comment:
perhaps should cause some sideeffect that we can check in the test - so we could know if something had to run in this context, it indeed has run in it. 
Maybe incrementing a counter in the context instance would be enough.